File: 1573-find-two-non-overlapping-sub-arrays-each-with-target-sum/find-two-non-overlapping-sub-arrays-each-with-target-sum.java
class Solution {
    public int minSumOfLengths(int[] arr, int target) {
        int n = arr.length;
        int[] best = new int[n];
        Arrays.fill(best, Integer.MAX_VALUE);
        int minSum = Integer.MAX_VALUE;
        int left = 0;
        int currentSum = 0;
        
        for (int right = 0; right < n; right++) {
            currentSum += arr[right];
            
            while (currentSum > target && left <= right) {
                currentSum -= arr[left];
                left++;
            }
            
            if (currentSum == target) {
                int l = right - left + 1;
                if (left > 0 && best[left - 1] != Integer.MAX_VALUE) {
                    minSum = Math.min(minSum, l + best[left - 1]);
                }
                best[right] = (right > 0) ? Math.min(best[right - 1], l) : l;
            } else {
                if (right > 0) {
                    best[right] = best[right - 1];
                }
            }
        }
        
        return minSum == Integer.MAX_VALUE ? -1 : minSum;
    }
}

File: 1573-find-two-non-overlapping-sub-arrays-each-with-target-sum/find-two-non-overlapping-sub-arrays-each-with-target-sum.py
class Solution:
    def minSumOfLengths(self, arr: list[int], target: int) -> int:
        n = len(arr)
        best = [float('inf')] * n
        min_sum = float('inf')
        left = 0
        current_sum = 0
        
        for right in range(n):
            current_sum += arr[right]
            
            while current_sum > target and left <= right:
                current_sum -= arr[left]
                left += 1
                
            if current_sum == target:
                l = right - left + 1
                if left > 0 and best[left - 1] != float('inf'):
                    min_sum = min(min_sum, l + best[left - 1])
                best[right] = min(best[right - 1] if right > 0 else float('inf'), l)
            else:
                if right > 0:
                    best[right] = best[right - 1]
                    
        return min_sum if min_sum != float('inf') else -1